File: sdk/python/autorca.py
import os
import sys
import traceback
import json
import requests
from typing import Optional, Dict, Any

class AutoRCAClient:
    """
    AutoRCA Python SDK & Middleware Wrapper
    Plug-and-play client to report runtime errors and trigger autonomous RCA swarm fixes.
    """

    def __init__(
        self,
        endpoint: str,
        api_key: Optional[str] = None,
        tenant_id: Optional[str] = None,
        project_id: Optional[str] = None,
        target_repo: Optional[str] = None,
        target_branch: Optional[str] = None,
        environment: Optional[str] = None
    ):
        self.endpoint = endpoint.rstrip("/")
        self.api_key = api_key or os.environ.get("AUTORCA_API_KEY")
        self.tenant_id = tenant_id or os.environ.get("AUTORCA_TENANT_ID", "org-acme-corp")
        self.project_id = project_id or os.environ.get("AUTORCA_PROJECT_ID", "proj-main")
        self.target_repo = target_repo or os.environ.get("AUTORCA_TARGET_REPO", "")
        self.target_branch = target_branch or os.environ.get("AUTORCA_TARGET_BRANCH", "main")
        self.environment = environment or os.environ.get("ENV", "production")

    def dispatch_investigation(
        self,
        title: str,
        error_message: str,
        stack_trace: Optional[str] = None,
        bug_id: Optional[str] = None,
        harness_command: str = "pytest",
        metadata: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Dispatch an automated RCA investigation job to the AutoRCA cluster.
        """
        url = f"{self.endpoint}/api/worktree/dispatch"
        payload = {
            "bugId": bug_id or f"PY-{int(os.getpid())}",
            "title": title,
            "errorMessage": error_message,
            "stackTrace": stack_trace or traceback.format_exc(),
            "repoUrl": self.target_repo,
            "branchName": self.target_branch,
            "tenantId": self.tenant_id,
            "projectId": self.project_id,
            "environment": self.environment,
            "harnessCommand": harness_command,
            "metadata": metadata or {},
        }

        headers = {
            "Content-Type": "application/json",
            "x-tenant-id": self.tenant_id,
            "x-project-id": self.project_id,
        }
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        try:
            response = requests.post(url, json=payload, headers=headers, timeout=5)
            response.raise_for_status()
            return response.json()
        except Exception as e:
            print(f"[AutoRCA Python SDK] Exception dispatch error: {e}", file=sys.stderr)
            return {"success": False, "error": str(e)}

    def capture_exception(self, exc: Exception, harness_command: str = "pytest", metadata: Optional[Dict[str, Any]] = None):
        """
        Convenience method to capture a caught Python exception.
        """
        tb = "".join(traceback.format_exception(type(exc), exc, exc.__traceback__))
        return self.dispatch_investigation(
            title=type(exc).__name__,
            error_message=str(exc),
            stack_trace=tb,
            harness_command=harness_command,
            metadata=metadata
        )

# FastAPI / Starlette Middleware Wrapper Example
def get_fastapi_middleware(client: AutoRCAClient):
    from starlette.middleware.base import BaseHTTPMiddleware

    class AutoRCAMiddleware(BaseHTTPMiddleware):
        async def dispatch(self, request, call_next):
            try:
                return await call_next(request)
            except Exception as exc:
                client.capture_exception(
                    exc,
                    metadata={"url": str(request.url), "method": request.method}
                )
                raise exc

    return AutoRCAMiddleware

File: sdk/ruby/autorca.rb
# frozen_string_literal: true

require 'net/http'
require 'json'
require 'uri'

module AutoRCA
  class Client
    attr_reader :endpoint, :api_key, :tenant_id, :project_id, :target_repo, :target_branch

    def initialize(endpoint:, api_key: nil, tenant_id: 'org-acme-corp', project_id: 'proj-main', target_repo: '', target_branch: 'main')
      @endpoint = endpoint.chomp('/')
      @api_key = api_key || ENV['AUTORCA_API_KEY']
      @tenant_id = tenant_id || ENV['AUTORCA_TENANT_ID'] || 'org-acme-corp'
      @project_id = project_id || ENV['AUTORCA_PROJECT_ID'] || 'proj-main'
      @target_repo = target_repo || ENV['AUTORCA_TARGET_REPO'] || ''
      @target_branch = target_branch || ENV['AUTORCA_TARGET_BRANCH'] || 'main'
    end

    def dispatch_investigation(title:, error_message:, stack_trace: nil, harness_command: 'bundle exec rake test', metadata: {})
      uri = URI.parse("#{@endpoint}/api/worktree/dispatch")
      payload = {
        bugId: "RUBY-#{Time.now.to_i}",
        title: title,
        errorMessage: error_message,
        stackTrace: stack_trace,
        repoUrl: @target_repo,
        branchName: @target_branch,
        tenantId: @tenant_id,
        projectId: @project_id,
        harnessCommand: harness_command,
        metadata: metadata
      }

      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == 'https')
      http.open_timeout = 5
      http.read_timeout = 5

      request = Net::HTTP::Post.new(uri.path, {
        'Content-Type' => 'application/json',
        'x-tenant-id' => @tenant_id,
        'x-project-id' => @project_id
      })
      request['Authorization'] = "Bearer #{@api_key}" if @api_key

      request.body = payload.to_json
      response = http.request(request)

      JSON.parse(response.body) rescue { 'success' => false, 'code' => response.code }
    rescue StandardError => e
      warn "[AutoRCA Ruby SDK] Failed to dispatch investigation: #{e.message}"
      { 'success' => false, 'error' => e.message }
    end
  end

  # Rack / Rails Exception Handler Middleware
  class RackMiddleware
    def initialize(app, options = {})
      @app = app
      @client = options[:client] || AutoRCA::Client.new(endpoint: ENV.fetch('AUTORCA_ENDPOINT', 'http://localhost:3000'))
    end

    def call(env)
      @app.call(env)
    rescue StandardError => exception
      Thread.new do
        @client.dispatch_investigation(
          title: exception.class.name,
          error_message: exception.message,
          stack_trace: exception.backtrace&.join("\n"),
          metadata: { path: env['PATH_INFO'], method: env['REQUEST_METHOD'] }
        )
      end
      raise exception
    end
  end
end
